### DISCUSSION TITLE
RFC: HTTP/2

### DISCUSSION THREAD

**Discussion by finleygn** [vercel/next.js#13689] [Help]
# Summary

Introduce an optional HTTP/2 server via a flag, built into next.js.

The main purpose of this RFC is to implement HTTP/2 for next.js with backward compatibility for HTTP/1.1.

Content of this RFC assumes the usage of Nodes core `http2` module.

# Context

Currently including a HTTP/2 server in next.js requires that you write your own `server.js` which disables automatic static optimization.

Performance benefits should be seen when using the HTTP/2 server because of:

- Multiplexing
- Header Compression
- A Single Connection
- Binary format
- Push methods **SEE DISCUSSION POINT 1**

Therefore implementing this as a core part of next will mean services running on next (over a secured connection) will see faster load times.

Next internally uses the `http` module, the `http2` module is what should be used to implement this because of nodes [Compatibility API](https://nodejs.org/docs/latest-v13.x/api/http2.html#http2_compatibility_api).

# Implementation

## CLI

Only a simple change to the CLI, adding a `--http2` flag to `start` and `dev` commands.

## Certifications

A certs folder located in the root of the project would contain a `key.pem` and `cert.pem`. 

The HTTP/2 spec does not state that the connection is required to be over TLS, however, modern browsers have not implemented HTTP/2 for any connections that are not secure. For this reason, the certs would be required in production and any other environment, including local.

**SEE DISCUSSION POINT 2**

# Notable Discussion Points

1. Push. One of the main benefits of using HTTP/2 is that it lets the server push resources to the client that they didn't initially request. Is this something we want to give the option to do?
2. Certs. For environments that provide a secure connection automatically, would we need local certs if the connection is secured upstream? Not sure how this works for something like Vercel's automatic SSL as well.

# Actions

- Create HTTP/2 flag
- Add a way to read certs
- Docs for HTTP/2, with info on how to create certs
- Refactor current server to be compatible with HTTP/2. Any modification to the req object isn't possible so middleware that does this needs to be modified. Any other incompatibilities need to be resolved

**Comment by Vadorequest** [comment]
I may be out of the subject here, but since Next.js is being built by Vercel and hosted on Vercel, how does it work on Vercel platform? Does it uses HTTP/1.1 or HTTP/2? (And if so, how?)

**Reply by finleygn** [reply]
that's one of my concerns, not sure how to handle hosting that provides SSL upstream or hosting that might not support HTTP/2 at all

node still offers a way to create an insecure HTTP/2 server if no certs can be given but need to look into the implications of that

**Comment by Manikanta-20** [comment]


**Reply by timneutkens** [reply]
Can you stop spamming random discussions and opening one line discussions, it comes over as behavior that a bot would have.

**Reply by Manikanta-20** [reply]
@timneutkens , we are discuss about http/2. I strike out https  also there it should no break.

**Comment by HugoDerigny** [comment]
What about now ? Has it implemented ?

**Comment by Apezdr** [comment]
With this going on the backburner, should the approach be to host your own server to serve NextJS? Some direction on how to mitigate this would be a good step.

**Comment by blodyle** [comment]
Nice! This would make it really easy!
Any news on this?
